File: 프로그래머스/4/60060. 가사 검색/가사 검색.py
class TrieNode:
    def __init__(self):
        self.count = 0
        self.childern = {}


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for w in word:
            if w not in node.childern:
                node.childern[w] = TrieNode()
            node = node.childern[w]
            node.count += 1

    def search(self, query):
        node = self.root
        for q in query:
            if q == '?':
                return node.count
            if q not in node.childern:
                return 0
            node = node.childern[q]
        return node.count


def solution(words, queries):
    trie = {}
    reversed_trie = {}
    answer = []

    # trie 초기화
    for word in words:
        length = len(word)
        if length not in trie:
            trie[length] = Trie()
            reversed_trie[length] = Trie()
        trie[length].insert(word)
        reversed_trie[length].insert(word[::-1])

    for query in queries:
        length = len(query)

        if length not in trie:
            answer.append(0)
            continue

        # 전체가 와일드 카드인 경우 처리
        if query[-1] == '?' and query[0] == '?':
            total = 0
            for i in trie[length].root.childern.values():
                total += i.count
            answer.append(total)
            continue

        # 접두사인 경우 reversed 적용
        if query[0] == '?':
            result = reversed_trie[length].search(query[::-1])
        else:
            result = trie[length].search(query)
        answer.append(result)

    return answer

File: 프로그래머스/4/60060. 가사 검색/가사 검색.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class TrieNode {
    public int count = 0;
    public HashMap<Character, TrieNode> children;

    public TrieNode() {
        this.count = 0;
        this.children = new HashMap<>();
    }
}

class Trie {
    TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word){
        TrieNode node = root;
        for(char c: word.toCharArray()){
            node.children.putIfAbsent(c, new TrieNode());
            node = node.children.get(c);
            node.count++;
        }
    }

    public int search(String query){
        TrieNode node = root;
        for (char q : query.toCharArray()) {
            if (q == '?'){
                return node.count;
            }
            if (!node.children.containsKey(q)){
                return 0;
            }
            node = node.children.get(q);
        }
        return node.count;
    }
}

class Solution {
    public List<Integer> solution(String[] words, String[] queries){
        Map<Integer, Trie> trie = new HashMap<>();
        Map<Integer, Trie> reversedTrie = new HashMap<>();
        List<Integer> answer = new ArrayList<>();

        for (String word : words){
            int length = word.length();
            trie.putIfAbsent(length, new Trie());
            reversedTrie.putIfAbsent(length, new Trie());

            trie.get(length).insert(word);
            reversedTrie.get(length).insert(new StringBuilder(word).reverse().toString());
        }

        for (String query : queries){
            int length = query.length();

            if (!trie.containsKey(length)){
                answer.add(0);
                continue;
            }

            if (query.charAt(length - 1) == '?' && query.charAt(0) == '?'){
                int total = 0;

                for (TrieNode t : trie.get(length).root.children.values()) {
                    total += t.count;
                }
                
                answer.add(total);
                continue;
            }
            int result;
            if (query.charAt(0) == '?'){
                result = reversedTrie.get(length).search(new StringBuilder(query).reverse().toString());
            }
            else {
                result = trie.get(length).search(query);
            }
            answer.add(result);
        }
        return answer;
    }
}
